Add check for ruby and rails versions

zheileman opened the pull request:
Please note, as per last commit message (6fa9a52), as it stands right now the check will result in a failure for non-Ruby or non-Rails repositories, so next step is to create a mechanism to bypass this check (and potentially others in the future) when not needed (i.e. not a Rails repo, so no need to check for Rails version).

In order to do this I kind of have an idea. We can have `pre-requisites` to be met in order for a check to run or not. A pre-requisite could be, for example, **uses_rails** and might encapsulate a little check in itself (i.e. could look for the existence of the `Gemfile`).

I didn't want to do this work as part of this PR as will potentially affect the BaseCheck and maybe Factory and Collection, and change quite a few tests. Also, I wanted to gather feedback first.

Of course, we can also have a **very pragmatic, interim solution**, which is to return a result of `not applicable` or something like that if none of the files we are looking for are found. After all, the pre-requisite for Rails would be to look for the gem in the Gemfile.lock, so it is implicit this file must exists in order to be able to check for such a gem version.

If we feel like not wanting to merge this PR without first having these `pre-requisites` or whatever mechanism we decide so we don't get false-negatives for non-ruby/rails repos, we can still comment out the version checks in the config file before merging it.

zheileman committed:
Declare the ruby version in the Gemfile

This seems to be the recommended way to declare the version of Ruby,
although it can also be specified in other files.

zheileman committed:
Add class to encapsulate version matching

This class encapsulate operations with versions, making use of the
already available `Gem::Requirement` and `Gem::Version` to create
a version requirement and validate it against a prospective version
that we've extracted from the provided `content` via a regex that
will be configurable in the config file.

We are making use of named regex captures (i.e. `version`) to make
it more explicit and easy to read when we use them in the config file.

zheileman committed:
Add check for ruby and rails versions

A new kind of Check has been added (`VersionCheck`) to provide us with a
simple framework to check the versioning of different things like in this
case `ruby` and `rails`.

The way it works is very similar to the `FileContentCheck` in which we
declare a list of files to check their content, and regex of the content
to check for, only in this case, these regex allow for capures of a
`version` to be checked against a requirement also declared in the config.

Currently it iterates through the provided list of files, and stops after
it finds the first one that meets the criteria.
If none found, it fails the check.

As it stands right now, it will fail for repositories without Rails (this
repository will fail this check) and also will fail for Python repos for
example, so next step is to create a mechanism to bypass this check if not
needed (i.e. not a Rails repo, so no need to check for Rails version).

digitalronin commented:
My feeling on this is that we should hide as much of the internal implementation as possible in the check class itself.
From the POV of the person writing the yaml file for their organisation, they should be able to say, "Tell me if the rails version is ~>4.2" or something like that.
The fact that the implementation of that check involves checking the Gemfile, .travis.yml or .ruby-version should be hidden from them. So, if some new packaging or ruby version management tool takes the ruby world by storm, people shouldn't have to go and update all of their repo-audit.yml files - upgrading to the latest version of the tool should be enough, because whoever is maintaining the 'rails version' check by that point should have done the work to add #{new tool} to the implementation.

SteveMarshall commented:
:+1: to what @digitalronin said. Though that does mean we then ought to do the same for any support languages… Which just makes me more inclined to consider checking that tools built for this purpose are set up, rather than doing it ourself?

digitalronin commented:
We're chatting about this verbally right now. @zheileman's point is that the "minimum version of a thing" check is very flexible, and the user can configure it to check the version of anything they want without repo-audit needing classes for 'check python version', 'check elixir version', 'check javascript version', ... 
One possible idea is to have some kind of 'macro' system in the config file, where 'check rails version' would pull in the standard library definition (which currently means 'check the Gemfile, then check ...')
So then the repo-audit maintainer would keep the standard definitions up to date, and users would get the benefit of that work just by updating the tool. But, the underlying implementation would be the 'check the version of a thing' check.

zheileman commented:
I get your points. Yes, I think since the beginning I've been trying to precisely abstract enough of the checks in configuration, so we can have generic `Check` classes that are reusable and independent of the underlying repo language (we have now 2 of these examples, one being the `FileContentCheck`, other being this new `VersionCheck`).

`VersionCheck` class is one, but in configuration could end up being repeated 4 or 5 times, each for every version we want to check, this can be ruby version, python version or even specific Gem versions (in fact, `rails` is in essence a check for a version of the Gem).

The macros idea is interesting and probably is starting to emerge as a pattern as we introduce more complex configuration.
The other option is to move to a 1:1 check architecture where a ruby check is a RubyVersionCheck, a rails check is a RailsVersionCheck, and so on... (not particularly attracted to this idea as in my head we might end having lots of literally identical classes only changing 1 regex, which would need to be declared in some constant, ...)

zheileman committed:
Add a constants module in first attempt to macros

Based on previous feedback, and realising the configuration file might
not be the correct place to hold very specific knowledge about ruby,
rails and in the future other languages by means of filenames and regex
to retrieve their versions, this commit extract these knowledge to a
`Constants` module and we get for free out of the box the capability of
injecting back into the configuration these details via ERB interpolation

It might be a very simple and rudimentary solution but might do for now
until we know how complex this can get.

SteveMarshall commented:
@digitalronin @zheileman How would you both feel about killing this in favour of checking https://dependabot.com (or similar) is configured? I feel like building our own version-checking stuff is a rocky road…

SteveMarshall commented:
@digitalronin @zheileman In fact, that's exactly what we were discussing in #19 . Are you both happy to kill this?

zheileman commented:
@SteveMarshall @digitalronin I've been a bit disconnected from this, but yep, that's fine with me. The interim solution in this PR was kind of a prototype, never intended to be a perfect solution or anything like that, just a low cost version check too coupled to regexes.

I think if Dependabot, or any other, are robust enough and they have some level of support behind them (as in: they are not the pet project of some kid who will no doubt dropping the project the moment they get bored, like SO MANY other projects out there) then it is realistic to not duplicate efforts 👍
